File: src/ffi/grovedb/src/checkpoint.cpp
// Copyright (c) 2026-present, The Dash Core developers
// Distributed under the MIT software license, see the accompanying
// file LICENSE.MIT or https://opensource.org/license/mit

#include <db_internal.h>
#include <util/ffi.h>

#include <rust/grovedb_cxx/lib.h>

#include <string_view>
#include <utility>

namespace grovedb {
// ---------------------------------------------------------------------------
// CreateCheckpoint
// ---------------------------------------------------------------------------

Result<void, Error> Db::CreateCheckpoint(std::string_view path)
{
  if (!m_impl) {
    return Err(Error::InvalidArgument("called on uninitialized database"));
  }
  return CallFFI([&]() -> Result<void, Error> {
    grovedb_cxx::grovedb_create_checkpoint(*m_impl->m_db, rust::Str(path.data(), path.size()));
    return {};
  });
}

// ---------------------------------------------------------------------------
// OpenCheckpoint
// ---------------------------------------------------------------------------

Result<Db, Error> Db::OpenCheckpoint(std::string_view path)
{
  return CallFFI([&]() -> Result<Db, Error> {
    auto boxed = grovedb_cxx::grovedb_open_checkpoint(rust::Str(path.data(), path.size()));
    Db db;
    db.m_impl = std::make_unique<Impl>(std::move(boxed));
    return db;
  });
}

// ---------------------------------------------------------------------------
// DeleteCheckpoint
// ---------------------------------------------------------------------------

Result<void, Error> Db::DeleteCheckpoint(std::string_view path)
{
  return CallFFI([&]() -> Result<void, Error> {
    grovedb_cxx::grovedb_delete_checkpoint(rust::Str(path.data(), path.size()));
    return {};
  });
}
} // namespace grovedb

File: rust/grovedb_cxx/src/checkpoint.rs
//
// Copyright (c) 2026-present, The Dash Core developers
// SPDX-License-Identifier: MIT
// See the accompanying file LICENSE or https://opensource.org/license/mit
//

use grovedb::GroveDb;

use crate::BoxedGroveDb;

// ---------------------------------------------------------------------------
// Checkpoint operations
// ---------------------------------------------------------------------------

/// Create a checkpoint of the database at the given filesystem path.
pub fn grovedb_create_checkpoint(db: &BoxedGroveDb, path: &str) -> Result<(), String> {
  db.db.create_checkpoint(path).map_err(crate::ffi_error)
}

/// Open an existing checkpoint as a read-only database handle.
pub fn grovedb_open_checkpoint(path: &str) -> Result<Box<BoxedGroveDb>, String> {
  let db = GroveDb::open_checkpoint(path).map_err(crate::ffi_error)?;
  Ok(Box::new(BoxedGroveDb { db }))
}

/// Delete a checkpoint directory.
///
/// Verifies the path is a valid checkpoint before deletion.
pub fn grovedb_delete_checkpoint(path: &str) -> Result<(), String> {
  GroveDb::delete_checkpoint(path).map_err(crate::ffi_error)
}
